### DISCUSSION TITLE
Sort a collection by taxonomy terms then show all sorted by date in a single loop?

### DISCUSSION THREAD

**Discussion by austriker27** [statamic/cms#6099] [Q&A]
Is there a way to sort a collection to show entries that include the same taxonomy term then show all sorted by date? (With a limit of 3)

Right now I've got a if conditional that does this separately but I reckon it's more performant to do it in a single loop. Perhaps this is possible with the new antlers parser (which I'm using but still very new to) where I could build an array locally using my desired filters and sorting? 

Here's my double loop with conditional (and `total_items` used to limit the return to 3 entries): 
```handlebars
{{ total_items = 0 }}
{{ collection:blog
    limit="{{ 3 | subtract:total_items }}"
    taxonomy:tags="{ tags }"
    :sort="date:desc"
    :id:not="id"
}}
    {{ if !no_results }}
    {{ total_items += 1 }}
        {{ partial:card }}
    {{ /if }}
{{ /collection:blog }}

{{ if total_items < 3 }}
{{ collection:blog
    taxonomy:tags:not="{ tags }"
    limit="{{ 3 | subtract:total_items }}"
    :sort="date:desc"
    :id:not="id"
}}
        {{ partial:card }}
{{ /collection:blog }}
{{ /if }}
```

I was thinking maybe something like this below but every variation I try doesn't work:

```handlebars
{{ collection:blog
  limit="3"
  :sort="taxonomy:tags|date:desc"
  :id:not="id"
}}
  {{ partial:card }}
{{ /collection:blog }}
```

**Comment by edalzell** [comment]
Why are you sorting twice, you said show all that have a term then sort by date?

**Reply by austriker27** [reply]
> Why are you sorting twice, you said show all that have a term then sort by date?

Because I want to first show any posts that contain the same taxonomy terms. Then I want to show any other posts sorted by date.

**Comment by JohnathonKoster** [comment]
A quick way to accomplish this would be to create a custom modifier that returns the number of items in common between the entry's terms and the page's terms (does not handle the case if any value is `null`):

```php
<?php

namespace App\Modifiers;

use Statamic\Modifiers\Modifier;

class IntersectionCount extends Modifier
{

    public function index($value, $params, $context)
    {
        // Assumes both the value and param will be an OrderedQueryBuilder
        // that contains a TermQueryBuilder inside.

        return collect($value->get()->all())->pluck('slug')
            ->intersect(collect($params[0]->get()->all())->pluck('slug'))
            ->count();
    }
}

```

We can then use `orderby` to sort the returned entries on two fields (the first will be the number of terms in common, and the second will be the date):

```antlers
{{ collection:blog :id:not="id" limit="3" sort="date:desc" as="entries" }}
    {{ blog_posts = entries orderby (
            (x => x.tags | intersection_count(page:tags)) 'desc',
            (x => x.date) 'desc') }}
        {{ title }}<br />
    {{ /blog_posts }}
{{ /collection:blog }}
```

> Note: This example is using the relatively rare `(x => x.date)` syntax. In this instance the `orderby` would receive the `Entry` objects itself, and the arrow syntax is an easy way to pull values out of an object. Currently will result in `array_merge` errors without it.

With the modifier and the ordering on two columns, all entries that have terms in common with the current page will appear first in the list. Any items that do not have terms in common will appear at the end, with a maximum of 3 total entries displayed.
